torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -1586,184 +1535,164 @@ Operation listList(const Node* node) {
   };
 }
 
-template <class TList, class TElement>
+template <class T>
 int listAdd(Stack& stack) {
-  TList a;
-  TList b;
+  c10::ListPtr<T> a = c10::make_list<T>();
+  c10::ListPtr<T> b = c10::make_list<T>();
   pop(stack, a, b);
 
-  std::vector<TElement> ret;
-  const auto total_size = a->elements().size() + b->elements().size();
+  c10::ListPtr<T> ret = c10::make_list<T>();
+  const auto total_size = a.size() + b.size();
   ret.reserve(total_size);
-  for (const auto& a_element : a->elements()) {
-    ret.push_back(a_element);
+  for (T a_element : a) {
+    ret.push_back(std::move(a_element));
   }
-  for (const auto& b_element : b->elements()) {
-    ret.push_back(b_element);
+  for (T b_element : b) {
+    ret.push_back(std::move(b_element));
```

Comment:
This is a functional change that I don't think is correct without checking that the list being moved from has only a single use.  See feedback from David and Michael on https://github.com/pytorch/pytorch/pull/21690

Reply:
This isn't moving from the list. It's copying from the list into `b_element` and then moving from `b_element`.

Reply:
out of curiosity, why was this code changed from the original way it was written?

I'd have expected something like
```
for (const T& b_element : b) {
  ret.push_back(b_element);
}
```

Reply:
Some `ListPtr<T>` (e.g. `ListPtr<string>` or nested lists/dicts) store their elements not as `T` but as IValues. When iterating over them, they have to create instances of `T` on the fly. This means we're copying `T`.

If the loop is `for (T e : list)`, then this instance is directly created as `e` and can be moved from there. If the loop is `for (const T& e : list)`, then the instance is also created as `e`, but now we can't move from it, so we need a second copy.